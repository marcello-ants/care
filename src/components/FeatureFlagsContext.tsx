import { createContext, useContext, ReactNode, useState } from 'react';
import { LDEvaluationDetail } from 'launchdarkly-node-server-sdk';

interface FeatureFlagsProviderProps {
  children: ReactNode;
  flags: FeatureFlags;
}

export type featureFlagsState = {
  flags: FeatureFlags;
};

export type FeatureFlags = Record<string, LDEvaluationDetail | undefined>;

const initialState: featureFlagsState = {
  flags: {},
};

const FeatureFlagsContext = createContext<featureFlagsState>(initialState);

export const FeatureFlagsProvider = ({ children, flags }: FeatureFlagsProviderProps) => {
  const [state] = useState({ flags });
  return <FeatureFlagsContext.Provider value={state}>{children}</FeatureFlagsContext.Provider>;
};

export const useFeatureFlags = () => useContext(FeatureFlagsContext);
