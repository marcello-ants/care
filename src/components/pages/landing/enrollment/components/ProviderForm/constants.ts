export const PROVIDER_SERVICES_DESKTOP = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const PROVIDER_SERVICES_MOBILE = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const DROPOFF_URLS = {
  CHILD_CARE: '/swf/ccOptimizedFlow',
  SENIOR_CARE: '/app/enrollment/provider/sc ',
  PET_CARE: '/swf/pcOptimizedFlow',
  HOUSEKEEPING: '/swf/hkOptimizedFlow',
  TUTORING: '/swf/tuOptimizedFlow',
  SPECIAL_NEEDS: '/swf/snOptimizedFlow',
  GENERIC: '/swf/cgOptimizedFlow',
  AU_PAIR: '',
  CARE_GIGS: '',
};
