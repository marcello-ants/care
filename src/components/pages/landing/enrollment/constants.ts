export const SITTER_FORM = {
  providerServices: {
    title: 'What Services do you offer?',
    id: 'providerService',
    options: [
      {
        value: 'CHILD_CARE',
        text: 'Babysitter',
      },
      {
        value: 'CHILD_CARE',
        text: 'Nanny',
      },
      {
        value: 'SPECIAL_NEEDS',
        text: 'Special Needs Provider',
      },
      {
        value: 'TUTORING',
        text: 'Tutor',
      },
      {
        value: 'TUTORING',
        text: 'Private Lesson Instructor',
      },
      {
        value: 'SENIOR_CARE',
        text: 'Senior Care Provider',
      },
      {
        value: 'PET_CARE',
        text: 'Pet Care Provider',
      },
      {
        value: 'HOUSEKEEPING',
        text: 'Housekeeper',
      },
      {
        value: 'GENERIC',
        text: 'Errands &amp; Odd Jobs Provider',
      },
    ],
  },
  providerEmail: {
    id: 'email',
    placeholder: 'Email',
  },
  providerPassword: {
    id: 'password',
    placeholder: 'Password',
  },
  providerZip: {
    id: 'providerZip',
    placeholder: 'ZIP',
  },
};

export const GENERIC_FORM = {
  howDidYouHearAboutUs: {
    title: 'How did you hear about us?',
    id: 'howDidYouHearAboutUs',
    options: [
      {
        value: 'FACEBOOK',
        text: 'Facebook',
      },
      {
        value: 'BANNER_AD',
        text: 'Banner Ad',
      },
      {
        value: 'TV_AD',
        text: 'TV Ad',
      },
      {
        value: 'OTHER_SOCIAL_MEDIA',
        text: 'Other Social Media (Twitter, Instagram, Pinterest, LinkedIn)',
      },
      {
        value: 'PRESS_COVERAGE',
        text: 'Press Coverage (News, Magazine, Blog)',
      },
      {
        value: 'FRIENDS_OR_FAMILY',
        text: 'Friends or Family',
      },
      {
        value: 'ONLINE_VIDEO',
        text: 'Online Video (Hulu, Youtube)',
      },
      {
        value: 'SEARCH_ENGINE',
        text: 'Search Engine (Google, Bing)',
      },
      {
        value: 'PARENTING_GROUP',
        text: 'Parenting Group or Forum',
      },
      {
        value: 'OTHER',
        text: 'Other',
      },
    ],
  },
};
