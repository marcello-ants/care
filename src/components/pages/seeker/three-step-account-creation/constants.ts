import * as yup from 'yup';
import { ServiceType } from '@/__generated__/globalTypes';
import { VerticalsAbbreviation } from '@/constants';

export const validationSchema = yup.object({
  firstName: yup
    .string()
    .min(2, 'First name needs to have at least two letters.')
    .required('First name is required'),
  lastName: yup
    .string()
    .min(2, 'Last name needs to have at least two letters.')
    .required('Last name is required'),
});

export const SERVICE_TYPE_BY_VERTICAL: { [key: string]: ServiceType } = {
  HK: ServiceType.HOUSEKEEPING,
  PC: ServiceType.PET_CARE,
  TU: ServiceType.TUTORING,
  CC: ServiceType.CHILD_CARE,
  DC: ServiceType.CHILD_CARE,
  IB: ServiceType.CHILD_CARE,
  SC: ServiceType.SENIOR_CARE,
};

export const ERROR_CREATING_ACCOUNT = 'An error occurred creating your account, please try again.';

export const FACEBOOK_VERTICALS_ALLOWANCE = ['CC'];
export const HDYHAU_VERTICALS_ALLOWANCE: VerticalsAbbreviation[] = ['HK', 'TU', 'PC', 'DC'];
