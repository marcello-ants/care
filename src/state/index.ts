/* eslint no-param-reassign: ["error", { "props": true, "ignorePropertyModificationsFor": ["draft"] }] */
import produce, { enableES5 } from 'immer';
import { CURRENT_STATE_VERSION, CURRENT_STATE_MAJOR_VERSION } from '../constants';
import { AppAction, AppState } from '../types/app';

import {
  SeekerAction,
  initialState as seekerInitialState,
  reducer as seekerReducer,
  validate as validateSeeker,
  doesSeekerStateHaveDefaultKeys,
} from './seeker';

import {
  SeekerCCAction,
  initialState as seekerCCInitialState,
  reducer as seekerCCReducer,
  validate as validateSeekerCC,
  doesSeekerCCStateHaveDefaultKeys,
} from './seekerCC';

import {
  ProviderAction,
  initialState as providerInitialState,
  reducer as providerReducer,
} from './provider';

import {
  ProviderCCAction,
  initialState as providerCCInitialState,
  reducer as providerCCReducer,
} from './providerCC';

import {
  SeekerHKAction,
  initialState as seekerHKInitialState,
  reducer as seekerHKReducer,
} from './seekerHK';

import {
  SeekerPCAction,
  initialState as seekerPCInitialState,
  reducer as seekerPCReducer,
} from './seekerPC';

import {
  SeekerTUAction,
  initialState as seekerTUInitialState,
  reducer as seekerTUReducer,
} from './seekerTU';

import { FlowAction, initialState as flowInitialState, reducer as flowReducer } from './flow';

import { LtcgAction, initialState as ltcgInitialState, reducer as ltcgReducer } from './ltcg';

// immer uses Proxy, but if that's not available the ES5 plugin needs to be enabled
/* istanbul ignore if */
if (typeof Proxy === 'undefined') {
  // thanks, IE 11
  enableES5();
}

export const initialAppState: AppState = {
  flow: flowInitialState,
  seeker: seekerInitialState,
  seekerCC: seekerCCInitialState,
  provider: providerInitialState,
  providerCC: providerCCInitialState,
  seekerHK: seekerHKInitialState,
  seekerPC: seekerPCInitialState,
  seekerTU: seekerTUInitialState,
  ltcg: ltcgInitialState,
  version: CURRENT_STATE_VERSION,
};

export const rootReducer = produce((draft: AppState, action: AppAction) => {
  draft.seeker = seekerReducer(draft.seeker, action as SeekerAction);
  draft.provider = providerReducer(draft.provider, action as ProviderAction);
  draft.providerCC = providerCCReducer(draft.providerCC, action as ProviderCCAction);
  draft.flow = flowReducer(draft.flow, action as FlowAction);
  draft.seekerCC = seekerCCReducer(draft.seekerCC, action as SeekerCCAction);
  draft.seekerHK = seekerHKReducer(draft.seekerHK, action as SeekerHKAction);
  draft.seekerPC = seekerPCReducer(draft.seekerPC, action as SeekerPCAction);
  draft.seekerTU = seekerTUReducer(draft.seekerTU, action as SeekerTUAction);
  draft.ltcg = ltcgReducer(draft.ltcg, action as LtcgAction);

  return draft;
});

export function validateState(state: AppState, pathName: string) {
  // TODO: call jobPost validation here instead of in AppState
  return validateSeeker(state, pathName) && validateSeekerCC(state, pathName);
}

export function isStateVersionCompatible(state: AppState) {
  if (state?.version) {
    const versionComponents = state.version.split('.');
    if (versionComponents.length === 3 && versionComponents[0] === CURRENT_STATE_MAJOR_VERSION) {
      return true;
    }
  }
  return false;
}

export function doesStateHaveDefaultKeys(state: AppState) {
  return (
    doesSeekerStateHaveDefaultKeys(state.seeker) && doesSeekerCCStateHaveDefaultKeys(state.seekerCC)
  );
}
