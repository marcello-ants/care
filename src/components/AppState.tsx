import { useRouter } from 'next/router';
import { createContext, useReducer, useContext, ReactNode, useEffect } from 'react';
import { debounce } from 'lodash-es';
import { configureScope, captureException, Severity } from '@sentry/nextjs';
import {
  initialAppState,
  rootReducer,
  isStateVersionCompatible,
  doesStateHaveDefaultKeys,
  validateState,
} from '@/state';
import patchAppState from '@/state/patch';
import { initialState as flowInitialState } from '@/state/flow';
import logger from '@/lib/clientLogger';
import { ClientLogger } from '@care/utils';
import getVertical from '@/utilities/verticalUtils';
import { SessionStorageHelper } from '@/utilities/SessionStorageHelper';
import { trackPageViewInSiftScience } from '@/utilities/siftScienceHelper';
import { AppAction, AppDispatch, AppState } from '@/types/app';
import { StateCheckingResponse } from '@/types/common';
import {
  LOCAL_STORAGE_STATE_KEY,
  CZEN_BASE_PATH,
  FLOWS,
  POST_A_JOB_ROUTES,
  CLIENT_FEATURE_FLAGS,
} from '@/constants';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

type AppStateProviderProps = {
  children: ReactNode;
  initialStateOverride?: AppState;
  czenJSessionId?: string | undefined;
  referrerCookie?: string | undefined | null;
  userHasAccount?: boolean;
  memberUuid?: string;
};

function getStateFromStorageOrDefault(defaultState: AppState) {
  let stateFromStorageJson;
  const sessionStorageAvailable = SessionStorageHelper.isAvailable();

  if (sessionStorageAvailable) {
    stateFromStorageJson = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
  } else {
    logger.warn({ event: 'sessionStorageUnavailable' });
  }

  if (stateFromStorageJson) {
    const stateFromStorage = JSON.parse(stateFromStorageJson) as AppState;
    try {
      const patchedState = patchAppState(stateFromStorage, defaultState);

      if (isStateVersionCompatible(patchedState) && doesStateHaveDefaultKeys(patchedState)) {
        return patchedState;
      }
      logger.warn({ event: 'invalidStateFromLocalStorage', state: stateFromStorageJson });
    } catch (e) {
      captureException(e, { level: Severity.Warning });
    }
  }
  return defaultState;
}

function persistState(state: AppState) {
  SessionStorageHelper.setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: true,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: AppState, action: AppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

export function redirectIfInvalidState(state: AppState, pathname: string): StateCheckingResponse {
  if (
    typeof Intl === 'undefined' ||
    typeof Intl.NumberFormat === 'undefined' ||
    typeof Intl.DateTimeFormat === 'undefined'
  ) {
    // TODO: Remove this with SC-214
    logger.error({ event: 'IntlUnavailable', source: pathname, invalidState: state });
    return {
      redirectCalled: true,
      readableMessage: 'Safari 9 and lower is not currently supported',
    };
  }

  const isStateValid = validateState(state, pathname);
  const redirect = !isStateValid;

  if (redirect) {
    SessionStorageHelper.removeItem(LOCAL_STORAGE_STATE_KEY);

    window.location.assign(CZEN_BASE_PATH);
  }

  return { redirectCalled: redirect };
}

export function getFlowName(pathname: string, oldFlowName?: string): string {
  let newFlowName = Object.values(FLOWS).find((flow) => pathname.includes(flow.path))?.name;

  if (!newFlowName) {
    // TODO: this shouldn't be needed after SC-123 is implemented
    const isPostAJob = Object.values(POST_A_JOB_ROUTES).some((route) => pathname.includes(route));
    if (isPostAJob) {
      newFlowName = 'SEEKER_POST_A_JOB';
    }
  }

  // Don't allow shifting from in-facility to seeker flow name
  if (newFlowName === FLOWS.SEEKER.name && oldFlowName === FLOWS.SEEKER_IN_FACILITY.name) {
    return FLOWS.SEEKER_IN_FACILITY.name;
  }

  return newFlowName ?? '';
}

function useFlowDetection(dispatch: AppDispatch, oldFlowName: string) {
  const { pathname } = useRouter();
  const newFlowName = getFlowName(pathname, oldFlowName);
  const vertical = getVertical();
  ClientLogger.setLoggerStatics({ flow: newFlowName, vertical });

  useEffect(() => {
    dispatch({ type: 'setFlowName', flowName: newFlowName });

    configureScope((scope) => {
      scope.setTags({
        flow: newFlowName,
        vertical,
      });
    });
  }, [newFlowName]);
}

const AppStateContext = createContext<AppState>(initialAppState);
const AppDispatchContext = createContext<AppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn(
    'Calling default dispatch is a no-op! You are probably consuming app dispatch in a component above AppStateProvider'
  )
);

export const AppStateProvider = ({
  children,
  initialStateOverride,
  czenJSessionId,
  referrerCookie,
  userHasAccount,
  memberUuid,
}: AppStateProviderProps) => {
  const hydratedInitialAppState = {
    ...initialAppState,
    flow: {
      ...flowInitialState,
      czenJSessionId,
      referrerCookie,
      userHasAccount,
      memberUuid,
    },
  };
  const featureFlags = useFeatureFlags();
  const sessionStorageRetrievalSuccessEvaluation =
    featureFlags?.flags[CLIENT_FEATURE_FLAGS.SESSION_STORAGE_RETRIEVAL_SUCCESS];
  const shouldMeasureSessionStorageRetrievalSuccess =
    sessionStorageRetrievalSuccessEvaluation?.variationIndex === 1;

  const router = useRouter();

  const logSessionStorageError = () => {
    logger.info({ event: 'sessionStorageStateUnavailable', path: router.pathname });
    AnalyticsHelper.logEvent({
      name: 'sessionStorageStateUnavailable',
    });
  };
  const launchDarklyWrappedGetStateFromStorageOrDefault = (defaultState: AppState) => {
    if (shouldMeasureSessionStorageRetrievalSuccess) {
      try {
        const sessionStorageState = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
        if (sessionStorageState && sessionStorageState.length > 0) {
          const stateFromStorage = JSON.parse(sessionStorageState);
          if (stateFromStorage) {
            logger.info({ event: 'sessionStorageStateAvailable', path: router.pathname });
          } else {
            logSessionStorageError();
          }
        } else {
          logSessionStorageError();
        }
      } catch (_e) {
        logSessionStorageError();
      }
    }
    return getStateFromStorageOrDefault(defaultState);
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    launchDarklyWrappedGetStateFromStorageOrDefault
  );
  useFlowDetection(dispatch, state.flow.flowName);
  const routePath = router.asPath?.split('?')[0];
  const stateCheckResult = redirectIfInvalidState(state, routePath);
  if (stateCheckResult.redirectCalled) {
    if (stateCheckResult.readableMessage) {
      return (
        <>
          <p>Redirecting to Member Home Page...</p>
          <p>{stateCheckResult.readableMessage}</p>
        </>
      );
    }
    return <></>;
  }

  useEffect(() => {
    trackPageViewInSiftScience(state);
  }, [routePath]);

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
};

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
  czenJSessionId: undefined,
  referrerCookie: undefined,
  userHasAccount: false,
  memberUuid: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);

export const useSeekerState = () => useContext(AppStateContext).seeker;
export const useProviderState = () => useContext(AppStateContext).provider;
export const useProviderCCState = () => useContext(AppStateContext).providerCC;
export const useFlowState = () => useContext(AppStateContext).flow;
export const useSeekerCCState = () => useContext(AppStateContext).seekerCC;
export const useSeekerHKState = () => useContext(AppStateContext).seekerHK;
export const useSeekerPCState = () => useContext(AppStateContext).seekerPC;
export const useSeekerTUState = () => useContext(AppStateContext).seekerTU;
export const useLtcgState = () => useContext(AppStateContext).ltcg;
