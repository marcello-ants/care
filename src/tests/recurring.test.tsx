import React from 'react';
import { useRouter } from 'next/router';
import { fireEvent, render, RenderResult, screen, waitFor } from '@testing-library/react';

import { initialAppState } from '@/state';
import { SENIOR_CARE_TYPE } from '@/__generated__/globalTypes';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import AmpliHelper from '@/utilities/ampliAnalyticsHelper';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';
import { CLIENT_FEATURE_FLAGS, POST_A_JOB_ROUTES } from '../constants';
import { AppStateProvider } from '../components/AppState';
import Page from '../pages/recurring';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

jest.mock('@/components/FeatureFlagsContext', () => ({
  useFeatureFlags: jest.fn().mockReturnValue({ flags: {} }),
}));

let mockRouter: any | null = null;
let renderResult: RenderResult;

const weekDays = [/Sun/i, /Mon/i, /Tue/i, /Wed/i, /Thu/i, /Fri/i, /Sat/i];
const careTimes = [/morning/i, /afternoon/i, /evening/i, /overnight/i];

const initialStateOverride: typeof initialAppState = {
  ...initialAppState,
  seeker: {
    ...initialAppState.seeker,
    typeOfCare: SENIOR_CARE_TYPE.IN_HOME,
  },
};

beforeEach(() => {
  mockRouter = {
    push: jest.fn(), // the component uses `router.push` only
    pathname: '',
    asPath: '',
  };
  (useRouter as jest.Mock).mockReturnValue(mockRouter);

  renderResult = render(
    <AppStateProvider initialStateOverride={initialStateOverride}>
      <Page />
    </AppStateProvider>,
    {}
  );
});

afterEach(() => {
  // cleanup on exiting
  mockRouter = null;
});

describe('Recurring page', () => {
  it('matches snapshot', () => {
    expect(renderResult.asFragment()).toMatchSnapshot();
  });
  it('should show an error message when the next button is pressed without a day selection', async () => {
    const morning = careTimes[0];
    screen.getByDisplayValue(morning).click();
    screen.getByRole('button', { name: 'Next' }).click();
    await waitFor(() =>
      expect(screen.getByText('Please select at least 1 day from the list below')).toBeVisible()
    );
    expect(mockRouter.push).not.toHaveBeenCalled();
  });
  it('should remove the day error message when a day is selected', async () => {
    screen.getByRole('button', { name: 'Next' }).click();
    await waitFor(() =>
      expect(screen.getByText('Please select at least 1 day from the list below')).toBeVisible()
    );
    const sunday = weekDays[0];
    screen.getByText(sunday).click();
    expect(
      screen.queryByText('Please select at least 1 day from the list below')
    ).not.toBeInTheDocument();
  });
  it('should show an error message when the next button is pressed without a time selection', async () => {
    const sunday = weekDays[0];
    screen.getByDisplayValue(sunday).click();
    screen.getByRole('button', { name: 'Next' }).click();
    await waitFor(() =>
      expect(screen.getByText('Please select at least 1 time from the list below')).toBeVisible()
    );
    expect(mockRouter.push).not.toHaveBeenCalled();
  });
  it('should remove the time error message when a day is selected', async () => {
    const morning = careTimes[0];
    const button = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(button);
    await waitFor(() =>
      expect(screen.getByText('Please select at least 1 time from the list below')).toBeVisible()
    );
    fireEvent.click(screen.getByText(morning));
    expect(
      screen.queryByText('Please select at least 1 time from the list below')
    ).not.toBeInTheDocument();
  });
  it('should remove the time error message when specific times mode is selected', async () => {
    screen.getByRole('button', { name: 'Next' }).click();
    await waitFor(() =>
      expect(screen.getByText('Please select at least 1 time from the list below')).toBeVisible()
    );
    fireEvent.click(screen.getByRole('button', { name: 'Add specific times instead' }));
    await waitFor(() =>
      expect(
        screen.queryByText('Please select at least 1 time from the list below')
      ).not.toBeInTheDocument()
    );
  });
  it('should display the day error message when specific times mode is selected with no days selected', async () => {
    screen.getByRole('button', { name: 'Add specific times instead' }).click();
    await waitFor(() =>
      expect(screen.getByText('Please select at least 1 day from the list below')).toBeVisible()
    );
  });
  it('should not show the "Times" header when specific times mode is selected with no days selected', () => {
    screen.getByRole('button', { name: 'Add specific times instead' }).click();
    expect(screen.queryByText('Times')).not.toBeInTheDocument();
  });
  it('routes to the pay-for-care page after selection and button click', async () => {
    const sunday = weekDays[0];
    const morning = careTimes[0];

    fireEvent.click(screen.getByText(sunday));
    fireEvent.click(screen.getByText(morning));
    fireEvent.click(screen.getByRole('button', { name: 'Next' }));
    await waitFor(() =>
      expect(mockRouter.push).toHaveBeenCalledWith(POST_A_JOB_ROUTES.PAY_FOR_CARE)
    );
  });

  it('fires both analytics events when feature flag is on', () => {
    const ampliListener = jest.spyOn(AmpliHelper.ampli, 'jobPostedSchedule');
    const amplitudeListener = jest.spyOn(AnalyticsHelper, 'logEvent');
    const sunday = weekDays[0];
    const morning = careTimes[0];

    (useFeatureFlags as jest.Mock).mockReturnValue({
      flags: {
        [CLIENT_FEATURE_FLAGS.AMPLITUDE_USE_AMPLI]: {
          variationIndex: 1,
          value: 'variation',
          reason: { kind: 'FALLTHROUGH' },
        },
      },
    });

    fireEvent.click(screen.getByText(sunday));
    fireEvent.click(screen.getByText(morning));
    fireEvent.click(screen.getByRole('button', { name: 'Next' }));

    expect(ampliListener).toBeCalledTimes(1);
    expect(amplitudeListener).toBeCalledTimes(1);
  });
});
