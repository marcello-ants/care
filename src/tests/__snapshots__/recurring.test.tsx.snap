// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recurring page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        class="MuiTypography-root makeStyles-header-5 MuiTypography-h2"
      >
        Which days do you need care?
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h4
        class="MuiTypography-root makeStyles-daySubheader-3 MuiTypography-h4"
      >
        Days
      </h4>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <ul
        class="MuiList-root makeStyles-rootInline-9 makeStyles-root-8 makeStyles-careDays-7 MuiList-padding"
      >
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.0"
            name="careDays"
            readonly=""
            type="checkbox"
            value="sun"
          />
          <div
            class="MuiBox-root MuiBox-root-21 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Sun
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.1"
            name="careDays"
            readonly=""
            type="checkbox"
            value="mon"
          />
          <div
            class="MuiBox-root MuiBox-root-36 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Mon
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.2"
            name="careDays"
            readonly=""
            type="checkbox"
            value="tue"
          />
          <div
            class="MuiBox-root MuiBox-root-37 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Tue
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.3"
            name="careDays"
            readonly=""
            type="checkbox"
            value="wed"
          />
          <div
            class="MuiBox-root MuiBox-root-38 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Wed
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.4"
            name="careDays"
            readonly=""
            type="checkbox"
            value="thu"
          />
          <div
            class="MuiBox-root MuiBox-root-39 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Thu
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.5"
            name="careDays"
            readonly=""
            type="checkbox"
            value="fri"
          />
          <div
            class="MuiBox-root MuiBox-root-40 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Fri
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careDays.6"
            name="careDays"
            readonly=""
            type="checkbox"
            value="sat"
          />
          <div
            class="MuiBox-root MuiBox-root-41 makeStyles-controlIndicator-14 xs"
            tabindex="0"
          >
            <p
              class="MuiTypography-root makeStyles-body3-25 MuiTypography-body1"
            >
              Sat
            </p>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h4
        class="MuiTypography-root makeStyles-timeSubheader-4 MuiTypography-h4"
      >
        Times
      </h4>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <ul
        class="MuiList-root makeStyles-rootInline-9 makeStyles-careTimes-43 MuiList-padding"
      >
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careTimes.0"
            name="careTimes"
            readonly=""
            type="checkbox"
            value="morning"
          />
          <div
            class="MuiBox-root MuiBox-root-46 makeStyles-controlIndicator-14 sm"
            tabindex="0"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Mornings
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careTimes.1"
            name="careTimes"
            readonly=""
            type="checkbox"
            value="afternoon"
          />
          <div
            class="MuiBox-root MuiBox-root-47 makeStyles-controlIndicator-14 sm"
            tabindex="0"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Afternoons
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careTimes.2"
            name="careTimes"
            readonly=""
            type="checkbox"
            value="evening"
          />
          <div
            class="MuiBox-root MuiBox-root-48 makeStyles-controlIndicator-14 sm"
            tabindex="0"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Evenings
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-root-10 pill MuiListItem-gutters"
        >
          <input
            class="makeStyles-input-13"
            id="selector.careTimes.3"
            name="careTimes"
            readonly=""
            type="checkbox"
            value="overnight"
          />
          <div
            class="MuiBox-root MuiBox-root-49 makeStyles-controlIndicator-14 sm"
            tabindex="0"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Overnight
            </p>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-specificTimesLink-44 MuiLink-button MuiTypography-body2 MuiTypography-colorPrimary"
      >
        Add specific times instead
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h6
        class="MuiTypography-root makeStyles-adjustLaterText-45 MuiTypography-subtitle2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        Don't worry, you can always adjust these times later.
      </h6>
    </div>
    <div
      class="MuiGrid-root makeStyles-nextButtonContainer-2 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Next
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
